Exiled.API/Features/Audio/AudioPlayer.cs

```diff
@@ -0,0 +1,405 @@
+// -----------------------------------------------------------------------
+// <copyright file="AudioPlayer.cs" company="Exiled Team">
+// Copyright (c) Exiled Team. All rights reserved.
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.API.Features.Audio
+{
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+
+    using Exiled.API.Features.Attributes;
+    using Exiled.API.Features.Audio.EventArgs;
+    using Exiled.API.Features.Core;
+    using Exiled.API.Features.DynamicEvents;
+    using MEC;
+    using Mirror;
+    using NVorbis;
+    using UnityEngine;
+    using VoiceChat;
+    using VoiceChat.Codec;
+    using VoiceChat.Codec.Enums;
+    using VoiceChat.Networking;
+
+    /// <summary>
+    /// Represents a NPC that can play audios.
+    /// </summary>
+    public class AudioPlayer : GameEntity
+    {
+        private readonly Queue<float> streamBuffer = new();
+
+        private CoroutineHandle playbackHandler;
+
+        private CoroutineHandle playbackUpdater;
+
+        private MemoryStream memoryStream;
+
+        private VorbisReader vorbisReader;
+
+        private float allowedSamples;
+
+        private int samplesPerSecond;
+
+        private float[] sendBuffer;
+
+        private float[] readBuffer;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AudioPlayer"/> class.
+        /// </summary>
+        /// <param name="npc">The npc that will be able to play audio.</param>
+        private AudioPlayer(Npc npc)
+            : base(npc.GameObject)
+        {
+            Owner = npc;
+            Dictionary.Add(npc, this);
+        }
+
+        /// <summary>
+        /// Gets all the audio players present in the server in the form of dictionary.
+        /// </summary>
+        public static Dictionary<Npc, AudioPlayer> Dictionary { get; } = new();
+
+        /// <summary>
+        /// Gets all the audio players present in the server in the form of list.
+        /// </summary>
+        public static new IEnumerable<AudioPlayer> List => Dictionary.Values.ToList();
+
+        /// <summary>
+        /// Gets or sets the <see cref="TDynamicEventDispatcher{T}"/> which handles all delegates to be fired before selecting an audio file.
+        /// </summary>
+        [DynamicEventDispatcher]
+        public static TDynamicEventDispatcher<SelectingAudioEventArgs> SelectingAudio { get; set; }
```

Comment:
Consider do `= new()` all dynamic events, because you're not initializing them with `DynamicEventManager::CreateFromTypeInstance(this)` anywhere, means they're going to generate a NRE.